### PR TITLE
Improve negotiate and kerberos packages

### PR DESCRIPTION
**Pull request by fifthdegree**
The goal of this pr is to tidy up and improve the kerberos package and allow for better transition between packages
specifically:
- Move the logic for acquiring kerberos credentials to the AcquireCredentialsHandle function
- Negotiate lets the sub-packages do the work of acquiring credentials
- Allow the same identity structure be used for both packages

### CODE REVIEW

**Comment by hardening** [comment]
Hum, is it me or the patch also changes the behaviour of the kerberos cache management ? I'd say it's better because it checks if the principal matches the default user cache, and if not instead of overriding the user cache, it will work in memory. It's better only if we don't plan to do any further use of that cache (may not be the case if we wanna implement RDPEA, but not a show stopper).

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 1]
```diff
@@ -253,463 +138,376 @@ static krb5_error_code krb5_prompter(krb5_context context, void* data, const cha
...
 
-	memset(&creds, 0, sizeof(creds));
-
-	if ((ret = krb5_get_init_creds_opt_alloc(ctx, &options)))
+	if (domain)
 	{
-		krb5_log_error_message(ctx, ret, WLOG_ERROR, "error while allocating options");
-		goto cleanup;
+		/* Convert to uppercase */
+		for (char* p = domain; *p != 0; p++)
```
[the remark this answers is not in the document]
Comment (fifthdegree) [reply winpr/libwinpr/sspi/Kerberos/kerberos.c]: changed to use CharUpper

**Comment by fifthdegree** [comment]
@hardening I'm pretty sure that unless explicitly destroyed, a memory cache would remain available for any thread in a process

**Comment by akallabeth** [comment]
@freerdp-bot test

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 2]
```diff
@@ -253,463 +138,373 @@ static krb5_error_code krb5_prompter(krb5_context context, void* data, const cha
...
-
-	WLog_Print_tag(TAG, level, "%s [%s]", buffer, msg);
-	free(buffer);
-	krb5_free_error_message(ctx, msg);
+static SECURITY_STATUS SEC_ENTRY kerberos_AcquireCredentialsHandleW(
+    SEC_WCHAR* pszPrincipal, SEC_WCHAR* pszPackage, ULONG fCredentialUse, void* pvLogonID,
+    void* pAuthData, SEC_GET_KEY_FN pGetKeyFn, void* pvGetKeyArgument, PCredHandle phCredential,
+    PTimeStamp ptsExpiry)
+{
+	return SEC_E_OK;
```
Comment (hardening) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:168]: That shall rather be something like `SEC_E_NOT_IMPLEMENTED`...

**Pull request closed without merging**